Lesson 5

ignatbagaev opened the pull request.

ignatbagaev committed:
add acceptance tests for creating question

Installed capybara and launchy in order to be able to make acceptance tests. Wrote acceptance test for creating question. Also installed twitter-bootstrap-rails gem to make the views a little bit more sightly.

ignatbagaev committed:
user could access to the listing of all questions

ignatbagaev committed:
add acceptance test for creating an answer

Wrote acceptance test for scenario of creating answer. Removed uselles :new action from answers controller and specs to test it's behaviour. Also added some links to navigate between pages.

ignatbagaev committed:
authorization

Added authorization of User with gem Devise and wrote corresponding tests.

ignatbagaev committed:
authenticate to answer

Restricted ability to create answers for users which are not authorized

ignatbagaev committed:
authenticate to ask

Make user to be authenticated to ask something.

ignatbagaev committed:
delete answer

User could delete his answers

ignatbagaev committed:
delete question

Make user be able to delete his own questions

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,23 +1,33 @@
 class AnswersController < ApplicationController
   before_action :load_question
```
Где аутентификация?

ignatbagaev replied:
@vkurennov
аутентификацию поставил в application controller, т.к. здесь для любых действий она нужна, а в questions_controller просто делаю skip перед show и index

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,23 +1,33 @@
-      render 'new'
-    end
+  def destroy
+    @answer.destroy if @answer.user == current_user
```
1. Нужно сравнивать по id, чтобы не грузить пользователя из базы лишний раз.
2. Т.к. у тебя нет аутентификации, current_user всегда будет вовзращать nil
3. Т.к. проверка на авторство будет встречаться часто, рекомнедую сделать метод в модели User, чтобы можно было вызвать примерно так: `current_user.author_of?(@answer)`

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,23 +1,33 @@
 
   def answer_params
-    params.require(:answer).permit(:body)
+    params.require(:answer).permit(:body).merge(user: current_user)
```
strong_params нужны для того, чтобы задвать список допустимых параметров, а не для того, чтобы их добавлять. Поэтому вынеси добавление current_user в create

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,23 +1,33 @@
+  end
+
+  def error
+    redirect_to @question
```
Нельзя делать редирект при ошибке, т.к. в этом случае потераяется информация об ошибках и пропадет все, что вводил пользователь. Можно делать только render.

vkurennov commented on app/controllers/application_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!
```
Не нужно здесь этого писать, лучше явно в каждом контролере задавать аутентификацию. У тебя далеко не все контрололеры и экшены будут требовать аутентификации. Да и так гораздо понятнее, когда сразу в контроллере видишь, что есть аутентификация и на какие экшены.

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -7,18 +8,33 @@ def show
 
+  def destroy
+    @question = Question.find(params[:id])
+    if @question.user == current_user
```
Проверяй по id

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -7,18 +8,33 @@ def show
 
   def question_params
-    params.require(:question).permit(:title, :body)
+    params.require(:question).permit(:title, :body).merge(user: current_user)
```
Перенеси добавление юзера в create

vkurennov commented on app/views/answers/_answer.html.slim:
```diff
@@ -0,0 +1,6 @@
+  td= @question.answers.index(answer) + 1
+  td= simple_format answer.body
+  td
+    -if answer.user == current_user
```
Используй метод User#author_of, о котором я выше написал и сравнивай по id

vkurennov commented on app/views/questions/_answer.html.slim:
```diff
@@ -0,0 +1,6 @@
+  td= @question.answers.index(answer) + 1
+  td= simple_format answer.body
+  td
+    -if answer.user == current_user
```
Используй метод User#author_of, о котором я выше написал и сравнивай по id

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,47 @@
+          td= @question.answers.index(answer) + 1
+          td= simple_format answer.body
+          td
+            - if answer.user == current_user
```
Здесь возникает стандартная проблема N+1 запроса: при проверке на автора, мы загружаем answer.user из базы. Т.е. мы сделали один запрос, получили список ответов, скажем, их было 7, и для каждого ответа мы еще делаем по запросу, и того получаем 8 запросов. Решение: сравнивать по id, а не по объекатм.

vkurennov commented on spec/acceptance/answer_specs/delete_answer_spec.rb:
```diff
@@ -0,0 +1,37 @@
+    user.answers << answer
+    log_in(user)
+    visit question_path(question)
+    expect { click_on 'Delete'}.to change(Answer, :count).by(-1)
```
В acceptance-тестах мы можем проверять только то, что видит пользователь. Представь, что ты живому тестеру пишешь сценарий. Ты же не будешь писать "нажми кнопку удалить, зайди в базу и проверь, что кол-во изменилось на 1". Ты напишешь "нажми кнопку удалить и убедись, что на странице исчез ответ". Вот и тут надо также тестировать.

vkurennov commented on spec/acceptance/question_specs/create_question_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    fill_in "Title", with: 'My question'
+    fill_in "Body",  with: 'Description of my issue'
+    click_on 'Ask'
+    expect(page).to have_content 'Question successfuly created'
```
Нужно проверить, что на странице появилис заголовок и тело вопроса

vkurennov commented on spec/acceptance/question_specs/delete_question_spec.rb:
```diff
@@ -0,0 +1,39 @@
+
+    visit question_path(question)
+
+    expect { click_link 'Delete question'}.to change(Question, :count).by(-1)
```
Проверяй изменения на странице, а не в базе

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -2,50 +2,70 @@
-              .to change(Answer, :count).by 1 
+      it 'it associates answer with question and saves it to database' do
+        expect { post :create, question_id: question.id,
+                               user_id: @user.id,
```
user_id не передается в параметрах, а берется из current_user.

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -2,50 +2,70 @@
-        expect { post :create, question_id: question.id, answer: attributes_for(:answer) }
-              .to change(question.answers, :count).by 1 
+      it 'it associates answer with user and saves it to database' do
+        expect { post :create, question_id: question.id, user_id: @user.id, answer: attributes_for(:answer) }
```
Тут тоже не нужен user_id

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -2,50 +2,70 @@
 
       it 'should redirect to question page' do
-        post :create, question_id: question.id, answer: attributes_for(:answer)
+        post :create, question_id: question.id, user_id: @user.id, answer: attributes_for(:answer)
```
Тут тоже не нужен user_id

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -2,50 +2,70 @@
       it 'should not create new answer' do
-        expect { post :create, question_id: question.id,
+        expect { post :create, question_id: question.id, 
+                               user_id: @user.id, 
```
Тут тоже не нужен user_id

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -26,25 +14,89 @@
+    describe 'POST #create' do
+      context 'with valid attributes' do
+        it 'saves new question to database' do
+          expect { post :create, question: attributes_for(:question) }.to change(Question, :count).by 1
```
Тут нужно проверить,что вопрос связывается с залогиненным пользователем

vkurennov commented on spec/factories/answers.rb:
```diff
@@ -1,6 +1,7 @@
   factory :answer do
     body "MyText"
-    question nil
+    association :user
```
Если имя ассоциации и фабрики совпадает, то можно писать так:

```
factory :answer do
      body 'MyText'
      question
      user
end
```

vkurennov commented on spec/acceptance/question_specs/questions_reading_spec.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  context 'when user authorized and there are questions' do
+    given!(:questions) { create_list(:question, 5) }
```
Нужно в фабриках через sequence сделать чтобы все вопросы были уникальными, иначе смысла проверять список нет, т.к. раз вопросы все одинаковые, то тест пройдет, даже если на странице будет 1 вопрос, а не 5.

vkurennov commented:
Исправь замечания

ignatbagaev committed:
fixes according to comments

ignatbagaev commented on app/models/question.rb:
```diff
@@ -3,4 +3,8 @@ class Question < ActiveRecord::Base
+
+  def answers
+    Answer.where(question_id: self.id)
+  end
```
почти уверен, что так нельзя делать, но сделал это потому что выскакивала ошибка на Questions#show, там в конце подгружался вопрос, которого в действительности не существовало. перерыл [active_record_querying](http://guides.rubyonrails.org/active_record_querying.html) в поисках лучшего решения, но ничего не сработало
в какую стронону кинуть взгляд, чтоб прозреть?

vkurennov replied:
метод `persisted?` т.е. внутри цикла проверяй, есть ли такой ответ в базе или нет через этот метод

ignatbagaev committed:
fix the way of getting answers on question page

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,17 +1,16 @@
-  def new
-    @answer = Answer.new
+  def create
+    @answer = @question.answers.new answer_params.merge(user: current_user)
```
Ставь скобки при вызове методов и передаче аргемунтов - это улучшает чтение. Исключение - различные DSL типа RSpec или хелперов в видах вроде link_to, form_for и т.п.

vkurennov commented on app/helpers/answers_helper.rb:
```diff
@@ -1,2 +1,5 @@
 module AnswersHelper
+  def answers_of(question_id)
```
И что это такое и зачем? По-человечески, через ассоциации нельзя вытащить ответы? `question.answers` и не надо огород городить

vkurennov commented on app/models/user.rb:
```diff
@@ -0,0 +1,13 @@
+  has_many :questions
+
+  def author_of?(obj)
+    obj.user_id == self.id ? true : false
```
Зачем тут тренарный оператор, если сравнение и так дает true или false? Оставь только сравнение

vkurennov replied:
Где тест модели на этот метод?

ignatbagaev replied:
@vkurennov , да, я обратил на это внимание когда сам перечитывал код
на форуме в ветке по гиту задал вопрос, по этому поводу
является ли хорошей практикой делать коммиты с целью просто улучишь читаемость и избавться от ненужных кусокв кода без внесения изменений в функционал
пока никто не ответил

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,23 @@
+div.col-md-offset-2.col-md-8.col-sm-12
+  h5.text-center
+    strong Answers:
+  - answers_of(@question).each do |answer|
```
Чем тебя не устраивает вариант `@question.answers.each?`

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -26,25 +14,90 @@
+    end
+  end
+
+  context 'when user is not logged in' do 
```
Ты не совсем верно вкладываешь контекст и describe. Нужно в describe выносить различные методы и экшены, а уже внутри делать разные контексты.

vkurennov commented on spec/acceptance/question_specs/questions_reading_spec.rb:
```diff
@@ -0,0 +1,57 @@
+  scenario 'users visits show page of certain question' do
+    visit question_path(question)
+    expect(page).to have_content(question.title)
+    expect(page).to have_content(question.title)
```
Надо еще проверить список ответов на странице, это важнейшая составляющая сценария. И не забудь сделать все ответы уникальными через sequence в фабриках.

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,17 +1,16 @@
-      render 'new'
-    end
+  def destroy
+    @answer.destroy if @answer.user_id == current_user.id
```
Используй метод author_of

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -1,24 +1,46 @@
+  end
+
+  def delete_question
+    if @question.user_id == current_user.id
```
Используй метод author_of

vkurennov replied:
Ну и почему не заменил эту проверку на метод author_of?

ignatbagaev replied:
@vkurennov, пардон. делал правки будучи на работе и видимо недостаточно сосредоточился - упустил из виду. больше не повторится)

vkurennov commented:
Исправь замечания

ignatbagaev committed:
more fixes

Fixed the way of showing the list of answers on question page. Removed unnecessary chunks of code. Wrote tests for User #author_of?. Refatored tests to make them more readable.

ignatbagaev commented on spec/models/user_spec.rb:
```diff
@@ -6,6 +6,20 @@
+    expect(user.author_of?(question)).to eq true
+  end
+  
+  it 'returns false when object is associated with user' do
```
допустил здесь описку
вместо 'is not' написал 'is'
поправлю в следующем коммите

vkurennov replied:
Нашел ошибку - сделай коммит сразу. Я же не мгновенно буду проверять, так что пока до проверки дойдет, ты уже что-то поправишь

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,17 +1,16 @@
-  def new
-    @answer = Answer.new
+  def create
+    @answer = Answer.new(answer_params.merge(user: current_user, question: @question))
```
Тут надо совместить merge и ассоциации, а то сильно громоздко получается. Например, так:
`@question.answers.new(answer_params.merge(user: current_user))`

ignatbagaev replied:
@vkurennov 
вот, кстати, как раз из-за того,что у меня было написано по такому образцу, у меня была проблема на странице questions#show о которой я писал в отчете

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -1,24 +1,46 @@
-    else
-      render 'new'
-    end
+    @question = Question.create question_params.merge(user: current_user)
```
Скобки у create

vkurennov commented on spec/acceptance/question_specs/questions_reading_spec.rb:
```diff
@@ -0,0 +1,66 @@
+  end
+
+  given(:answers) {create_list(:answer, 5)} 
+  before { question.answers << answers }
```
Устанавливай ассоциации сразу в given/let:
`given(:answers) {create_list(:answer, 5, question: question)}`

vkurennov commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,25 @@
+  it { should respond_to(:author_of?)}
+
+  it 'returns true when object is associated with user' do
+    user = FactoryGirl.create(:user)
```
Выноси тестовые данные в let, у тебя же как минимум user и question в обоих тестах повторяются. И ассоциации ставь сразу в let.

vkurennov commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,25 @@
+    user = FactoryGirl.create(:user)
+    question = FactoryGirl.create(:question)
+    user.questions << question
+    expect(user.author_of?(question)).to eq true
```
Если у тебя есть метод-предикат (заканчивается на знак вопроса и возвращает true/false), то rspec умеет делать красивые метчеры для них на основе имени:

`expect(user).to be_author_of(question)`

Т.е. имя метчера - это имя метода без вопр. знака и в начале подставляем be_

vkurennov commented:
Замечания остались не критичные, поэтому принимаю. НО! Все замечания исправь при выполнеии следующего задания.

ignatbagaev committed:
refactoring
